{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import random\n",
    "from numpy.random import normal\n",
    "from random import shuffle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Genome:\n",
    "    '''\n",
    "    A genome made by n genes and having p ploidy, where p = 2 is a diploid\n",
    "    genome. In the current version of the software, only p = 2 is allowed.\n",
    "\n",
    "    The genome is initialized randomly, each gene having a random dosage of\n",
    "    the alleles.\n",
    "\n",
    "    By default, a genome is initialized without the ability to sexually\n",
    "    reproduce\n",
    "    '''\n",
    "    def __init__(self, n: int, p: int):\n",
    "        self.n_genes = n\n",
    "        self.ploidy = p\n",
    "        self.genes = [[random() < 0.5 for i in range(n)] for i in range(p)]\n",
    "        self.meiosis = False\n",
    "\n",
    "    def mutate(self, mutation_rate):\n",
    "        for i in range(len(self.genes)):\n",
    "            for j in range(len(self.genes[i])):\n",
    "                if random() < mutation_rate:\n",
    "                    self.genes[i][j] = not self.genes[i][j]\n",
    "\n",
    "    def generate_crossingover_points(self, h_mean: float, h_sd: float):\n",
    "        crossingover_points = []\n",
    "        x = 0\n",
    "        x = int(x + normal(loc=h_mean,\n",
    "                           scale=h_sd))\n",
    "        while x < self.n_genes:\n",
    "            crossingover_points.append(x)\n",
    "            x = int(x + normal(loc=h_mean,\n",
    "                               scale=h_sd))\n",
    "        if len(crossingover_points) == 0:\n",
    "            print(\n",
    "            f'''\n",
    "            W: no crossing-over points generated. Check the following variables:\n",
    "                * mean number of genes between consecutive crossing over joints: {h_mean}\n",
    "                * Std dev of the number of genes between consecutive crossing over joints: {h_sd}\n",
    "                * Number of genes in the chromosome: {self.n_genes}\n",
    "            ''')\n",
    "        return(crossingover_points)\n",
    "\n",
    "    def perform_crossingover(self, crossingover_points):\n",
    "        new_genes = self.genes\n",
    "        crossingover_points.insert(0, 0)\n",
    "        crossingover_points.insert(len(crossingover_points), self.n_genes - 1)\n",
    "        crossing_over_segments = list(zip(crossingover_points[:len(crossingover_points) -1], crossingover_points[1:]))\n",
    "        for s in range(len(crossing_over_segments)):\n",
    "            if s % 2 == 0:\n",
    "                start = crossing_over_segments[s][0]\n",
    "                end = crossing_over_segments[s][1] + 1\n",
    "                new_genes[0][start:end] = self.genes[1][start:end]\n",
    "                new_genes[1][start:end] = self.genes[0][start:end]\n",
    "        return(new_genes)\n",
    "\n",
    "    def generate_gametes(self, h_mean: float, h_sd: float):\n",
    "        if self.meiosis == False:\n",
    "            print('W: self.meiosis is False, but self.generate_gametes() was called anyway')\n",
    "        self.crossingover_length_mean = h_mean\n",
    "        self.crossingover_length_sd = h_sd\n",
    "        # duplicate genome\n",
    "        new_genome_a = [[] for i in range(self.ploidy)]\n",
    "        new_genome_b = [[] for i in range(self.ploidy)]\n",
    "        # determine crossing-over points separately for each one of the two copies\n",
    "        crossingover_points_a = self.generate_crossingover_points(h_mean = h_mean, h_sd = h_sd)\n",
    "        crossingover_points_b = self.generate_crossingover_points(h_mean = h_mean, h_sd = h_sd)\n",
    "        # perform crossing-over separately for each one of the two copies\n",
    "        new_genome_a = self.perform_crossingover(crossingover_points_a)\n",
    "        new_genome_b = self.perform_crossingover(crossingover_points_b)\n",
    "        # generate gametes (meiosis II)\n",
    "        self.gametes = [new_genome_a[0], new_genome_a[1], new_genome_b[0], new_genome_b[1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Simulation:\n",
    "    '''\n",
    "    An environment in which Genomes evolve.\n",
    "\n",
    "    Parameters:\n",
    "\n",
    "        * c: the carrying capacity, i.e. the maximum number of genomes that\n",
    "          can be concurrently sustained by the environment\n",
    "        * n: number of genes in the evolving Genomes\n",
    "        * p: ploidy of the evolving Genomes, where p = 2 is a diploid genome.\n",
    "             In the current version of the software, only p = 2 is allowed\n",
    "        * s: number of genomes that exist at the start of the simulation\n",
    "        * h_mean: mean number of genes between two crossing-over points\n",
    "          NOTE: the real number will be slightly lower due to the truncation\n",
    "          at the end of the chromosome\n",
    "        * h_sd: standard deviation of the number of genes between two\n",
    "          crossing-over points\n",
    "        * x: starting fraction of genomes that can sexually reproduce\n",
    "        * g: number of generation the simulation will run\n",
    "    \n",
    "    The following assumptions hold:\n",
    "\n",
    "        * The genome is constituted by only one chromosome\n",
    "        * Genes are ordered, meaning that a distance between two genes A and B\n",
    "          can be calculated in terms of how many genes separate A and B\n",
    "        * In the current version of the software, the genes' effects are\n",
    "          independent\n",
    "        * In the current version of the software, the effect of allele dosage\n",
    "          is additive\n",
    "        * Each gene has two alleles. One allele is supposed to have no effect\n",
    "          on the fitness, while the other can be beneficial or detrimental\n",
    "        * A genome is initialized randomly, each gene having a random dosage of\n",
    "          the alleles\n",
    "    '''\n",
    "    def __init__(self, c: int, n: int, p: int, s: int, h_mean: int, h_sd: float, x: float, g: int, m: float):\n",
    "        self.carrying_capacity = c\n",
    "        self.n_genes = n\n",
    "        self.ploidy = p\n",
    "        self.start_n_genomes = s\n",
    "        self.gene_effects = [random() - 0.5 for i in range(n)]\n",
    "        self.population = [Genome(n, p) for i in range(s)]\n",
    "        self.crossingover_length_mean = h_mean\n",
    "        self.crossingover_length_sd = h_sd\n",
    "        self.start_sex_frac = x\n",
    "        self.n_generations = g\n",
    "        self.mutation_rate = m\n",
    "\n",
    "    def propagate(self):\n",
    "        genomes_indexes = [i for i in range(len(self.population))]\n",
    "        fitness_list = [self.compute_fitness(self.population[i]) for i in range(len(self.population))]\n",
    "        propagation_list = [x for _, x in sorted(zip(fitness_list, genomes_indexes))]\n",
    "        new_population = []\n",
    "        while len(new_population) < self.carrying_capacity and len(propagation_list) > 0:\n",
    "            if self.population[propagation_list[0]].meiosis:\n",
    "                # 1. remove and return the first genome from the queue propagation_list\n",
    "                # 2. identify next in the propagation_list that has meiosis, at index j\n",
    "                # 3. if no j, move to next iteration\n",
    "                # 4. if j, remove and return it from propagation_list\n",
    "                parent_a = propagation_list.pop(0)\n",
    "                j = 0\n",
    "                while self.population[propagation_list[j]].meiosis == False and j < len(propagation_list):\n",
    "                    j = j + 1\n",
    "                if j >= len(propagation_list):\n",
    "                    parent_b = propagation_list.pop(j)\n",
    "                    # 5. perform meiosis in parent_a and parent_b\n",
    "                    parent_a.generate_gametes(h_mean = self.crossingover_length_mean,\n",
    "                                              h_sd = self.crossingover_length_sd)\n",
    "                    parent_b.generate_gametes(h_mean = self.crossingover_length_mean,\n",
    "                                              h_sd = self.crossingover_length_sd)\n",
    "                    # 6. generate 4 new genomes and append them in new_population\n",
    "                    shuffled_gametes_idx = [i for i in range(len(parent_a.gametes))]\n",
    "                    shuffle(shuffled_gametes_idx)\n",
    "                    for g in range(len(shuffled_gametes_idx)):\n",
    "                      new_genes = [parent_a.gametes[g], parent_b.gametes[shuffled_gametes_idx[g]]]\n",
    "                      new_genome = Genome(n=self.n_genes, p=self.ploidy)\n",
    "                      new_genome.meiosis = True\n",
    "                      new_genome.genes = new_genes\n",
    "                      new_population.append = new_genome                      \n",
    "            else:\n",
    "                new_population.append(self.population[propagation_list[i]]) # duplicate the individual\n",
    "                new_population.append(self.population[propagation_list[i]])\n",
    "            i = i + 1\n",
    "\n",
    "    def mutate_genomes(self):\n",
    "      for i in range(len(self.population)):\n",
    "        self.population[i].mutate(self.mutation_rate)\n",
    "\n",
    "    def start(self):\n",
    "      # initialize meiosis attribute in the population according to frequency x\n",
    "      for i in range(self.start_n_genomes):\n",
    "        if random() < self.start_sex_frac:\n",
    "          self.population[i].meiosis = True\n",
    "      # run simulation over n generations\n",
    "      for i in range(self.n_generations):\n",
    "        self.propagate()\n",
    "        self.mutate_genomes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    CARRYING_CAPACITY = 10000\n",
    "    START_N_GENOMES = 1000\n",
    "    N_GENES = 100 #10000\n",
    "    PLOIDY = 2\n",
    "    MUTATION_RATE = 5 / N_GENES\n",
    "    CROSSING_OVER = True\n",
    "    CROSSINGOVER_LENGTH_MEAN = 5 #50\n",
    "    CROSSINGOVER_LENGTH_SD = 1 #5\n",
    "    START_FRAC_SEX_GENOMES = 0.1\n",
    "    N_GENERATIONS = 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "f0210e761cedf0284dba26e7c1120df0c2fff86ec80306828223a760cb71dfcb"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('sex_emergence': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
